[coop] Add configuration option to disable stack scan

BrzVlad opened the pull request:
When this experiment is enabled, the GC doesn't scan stack or registers. This mode is meant to disable scanning of native frames. This implementation only makes sense in full interpreter mode, where the interp stack is scanned separately from the thread stack. Longer term we could use this mode to scan only the managed stack, also in jit mode.

The point of this mode is to be able to stress test the GC in a wasm like environment, where stack can't be scanned. We can more easily reproduce and investigate GC crashes in this configuration, that would also happen on wasm.

jaykrell commented on src/mono/mono/metadata/sgen-mono.c:
```diff
@@ -2361,6 +2361,7 @@ sgen_client_scan_thread_data (void *start_nursery, void *end_nursery, gboolean p
 
 		aligned_stack_start = get_aligned_stack_start (info);
 		g_assert (info->client_info.suspend_done);
+#ifndef ENABLE_EXPERIMENT_COOPROOTS
```
Can this be a global variable, controllable from command line or environment variable, instead of an ifdef?
Why should I have to rebuild an replace the runtime, just to add/remove a few lines?
Can the ifdef merely control the default value of a global variable instead? But still allow both options no matter the configuration? ifdef is over used.

naricc commented:
So is sgen_client_scan_thread_data only used when ENABLE_EXPERIMENT_COOPROOTS is 1? I am confused what use it is when all the code is ifdefed way?

vargaz commented:
It could also be a MONO_GC_DEBUG option. Enabling/disabling it using configure sounds painful.

BrzVlad commented:
The end goal of this change is to run this configuration on CI to see what failures we can catch and from what I understand it seemed easier to have a runtime that doesn't scan the stack, without any additional required args at run time.

@akoeplinger @safern Is it fine to do such a lane which runs all tests while having some environment variables set.

BrzVlad committed:
[coop] Add debug option to disable stack scan

When this is enabled, the GC doesn't scan stack or registers. This mode is meant to disable scanning of native frames. This implementation only makes sense in full interpreter mode, where the interp stack is scanned separately from the thread stack. Longer term we could use this mode to scan only the managed stack, also in jit mode.

The point of this mode is to be able to stress test the GC in a wasm like environment, where stack can't be scanned. We can more easily reproduce and investigate GC crashes in this configuration, that would also happen on wasm.

BrzVlad commented:
switched to controlling this option through MONO_GC_DEBUG
